### ISSUE TITLE
Incorrect error handling for invalid input names on SavedModel export

### ISSUE THREAD

**Issue by dwyatte** [tensorflow/tensorflow#44984]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Any
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.3.1
- Python version: 3.7.6

**Describe the current behavior**

```
import tensorflow as tf

A_FEATURE_NAME = 'a feature'
B_FEATURE_NAME = 'b_feature'

class MyModel(tf.keras.Model):

    def call(self, inputs):
        return inputs[A_FEATURE_NAME] + inputs[B_FEATURE_NAME]

inputs = {
    A_FEATURE_NAME: tf.constant([1.]),
    B_FEATURE_NAME: tf.constant([2.]),
}

model = MyModel()
outputs = model(inputs)

model.save('test')
```

Saving this model gives the following error message

```
ValueError: Got non-flat/non-unique argument names for SavedModel signature 'serving_default': more than one argument to '__inference_signature_wrapper_48' was named 'a feature'. Signatures have one Tensor per named input, so to have predictable names Python functions used to generate these signatures should avoid *args and Tensors in nested structures unless unique names are specified for each. Use tf.TensorSpec(..., name=...) to provide a name for a Tensor input.
```

The reason however is that `A_FEATURE_NAME` has an invalid name due to it containing whitespace, so it ends up not matching the placeholder name when tracing the graph for saving: https://github.com/tensorflow/tensorflow/blob/5d8b439eb41fb2373c5b5c5f33c51313cfd9e57f/tensorflow/python/saved_model/save.py#L494

**Describe the expected behavior**
TensorFlow checks for valid names elsewhere (such as when [declaring an op](https://github.com/tensorflow/tensorflow/blob/00d1dc082eb0fde33e5d8944e4f203248be335f7/tensorflow/python/framework/ops.py#L1804)), so the same logic should be used to return a more relevant error e.g., `ValueError: 'a feature' is not a valid input name`

**Comment by Saduf2019** [comment]
I ran the code shared and i am able to replicate the error reported, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/5bdb0ceb7ad7db8569bc9bd485888a64/untitled463.ipynb).

**Comment by bhack** [comment]
@dwyatte 
```python
import tensorflow as tf

A_FEATURE_NAME = 'a feature'
B_FEATURE_NAME = 'b_feature'

class MyModel(tf.keras.Model):

    def call(self, inputs):
        return inputs['A_FEATURE_NAME'] + inputs['B_FEATURE_NAME']

inputs = {
    'A_FEATURE_NAME': tf.constant([1.]),
    'B_FEATURE_NAME': tf.constant([2.]),
}

model = MyModel()
outputs = model(inputs)

model.save('test')

**Comment by dwyatte** [comment]
@bhack thanks for the workaround, but I still think the wrong error is raised here. `more than one argument to '__inference_signature_wrapper_48' was named 'a feature'` suggests that I am using the same the same op name twice and in fact, the source code mentions `# This should be unreachable, since concrete functions may not be generated with non-unique argument names`. Instead, it might be nicer if tensorflow checked for valid op names before the aforementioned error is reached in `tf.keras.Model.save` and raised a more informative error (something like `ValueError: 'a feature' is not a valid input name`).

**Comment by bhack** [comment]
@dwyatte The problem with your issue is that invalid name are replaced to "

**Comment by dwyatte** [comment]
@bhack Thanks for the heads up, I was not aware there were changes in master that might handle address this. With `'2.5.0-dev20201127'` I do see the warning you mention, but I still receive an error which suggests the renaming is not happening:

```
...
WARNING:absl:Function `_wrapped_model` contains input name(s) a feature with unsupported characters which will be renamed to placeholder in the SavedModel.
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-bc463538c686> in <module>
     17 outputs = model(inputs)
     18 
---> 19 model.save('test')

~/venv/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py in save(self, filepath, overwrite, include_optimizer, save_format, signatures, options, save_traces)
   2046     # pylint: enable=line-too-long
   2047     save.save_model(self, filepath, overwrite, include_optimizer, save_format,
-> 2048                     signatures, options, save_traces)
   2049 
   2050   def save_weights(self,

~/venv/lib/python3.7/site-packages/tensorflow/python/keras/saving/save.py in save_model(model, filepath, overwrite, include_optimizer, save_format, signatures, options, save_traces)
    156   else:
    157     saved_model_save.save(model, filepath, overwrite, include_optimizer,
--> 158                           signatures, options, save_traces)
    159 
    160 

~/venv/lib/python3.7/site-packages/tensorflow/python/keras/saving/saved_model/save.py in save(model, filepath, overwrite, include_optimizer, signatures, options, save_traces)
     94       with utils.keras_option_scope(save_traces):
     95         saved_nodes, node_paths = save_lib.save_and_return_nodes(
---> 96             model, filepath, signatures, options)
     97 
     98     # Save all metadata to a separate file in the SavedModel directory.

~/venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py in save_and_return_nodes(obj, export_dir, signatures, options, raise_metadata_warning)
   1077   _, exported_graph, object_saver, asset_info, saved_nodes, node_paths = (
   1078       _build_meta_graph(obj, signatures, options, meta_graph_def,
-> 1079                         raise_metadata_warning))
   1080   saved_model.saved_model_schema_version = constants.SAVED_MODEL_SCHEMA_VERSION
   1081 

~/venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py in _build_meta_graph(obj, signatures, options, meta_graph_def, raise_metadata_warning)
   1263   with save_context.save_context(options):
   1264     return _build_meta_graph_impl(obj, signatures, options, meta_graph_def,
-> 1265                                   raise_metadata_warning)

~/venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py in _build_meta_graph_impl(obj, signatures, options, meta_graph_def, raise_metadata_warning)
   1197   asset_info, exported_graph = _fill_meta_graph_def(meta_graph_def,
   1198                                                     saveable_view, signatures,
-> 1199                                                     options.namespace_whitelist)
   1200   if options.function_aliases:
   1201     function_aliases = meta_graph_def.meta_info_def.function_aliases

~/venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py in _fill_meta_graph_def(meta_graph_def, saveable_view, signature_functions, namespace_whitelist)
    699 
    700   with exported_graph.as_default():
--> 701     signatures = _generate_signatures(signature_functions, resource_map)
    702     for concrete_function in saveable_view.concrete_functions:
    703       concrete_function.add_to_graph()

~/venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py in _generate_signatures(signature_functions, resource_map)
    563     mapped_inputs, exterior_argument_placeholders = (
    564         _map_function_arguments_to_created_inputs(argument_inputs,
--> 565                                                   signature_key, function.name))
    566     outputs = _call_function_with_mapped_captures(
    567         function, mapped_inputs, resource_map)

~/venv/lib/python3.7/site-packages/tensorflow/python/saved_model/save.py in _map_function_arguments_to_created_inputs(function_arguments, signature_key, function_name)
    504            "tf.TensorSpec(..., name=...) to provide a name for a Tensor "
    505            "input.").format(signature_key, compat.as_str_any(function_name),
--> 506                             user_input_name))
    507     arg_placeholder = array_ops.placeholder(
    508         shape=placeholder.shape,

ValueError: Got non-flat/non-unique argument names for SavedModel signature 'serving_default': more than one argument to '__inference_signature_wrapper_42' was named 'a feature'. Signatures have one Tensor per named input, so to have predictable names Python functions used to generate these signatures should avoid *args and Tensors in nested structures unless unique names are specified for each. Use tf.TensorSpec(..., name=...) to provide a name for a Tensor input.
```

**Comment by bhack** [comment]
The problem is that you are going ti mismatch between `featuere a` and substituted name placeholder at

https://github.com/tensorflow/tensorflow/blob/21eb35ad7c4c95343c872f68b7f298b66afc9b0e/tensorflow/python/saved_model/save.py#L494

**Comment by dwyatte** [comment]
Reading the comment above this conditional
https://github.com/tensorflow/tensorflow/blob/21eb35ad7c4c95343c872f68b7f298b66afc9b0e/tensorflow/python/saved_model/save.py#L494 
makes me think it should somehow be rewritten to not capture cases where inputs are simply renamed on saving.
```
    # If the internal placeholders for a function have names which were
    # uniquified by TensorFlow, then a single user-specified argument name
    # must refer to multiple Tensors. The resulting signatures would be
    # confusing to call. Instead, we throw an exception telling the user to
    # specify explicit names.
```

If I execute
https://github.com/tensorflow/tensorflow/blob/21eb35ad7c4c95343c872f68b7f298b66afc9b0e/tensorflow/python/saved_model/save.py#L507
I get the error I would expect:

```
arg_placeholder = array_ops.placeholder(shape=placeholder.shape, dtype=placeholder.dtype,name="{}_{}".format(signature_key, user_input_name))
*** ValueError: 'serving_default_a feature' is not a valid scope name
```

**Comment by bhack** [comment]
You could probably open a small PR adding an `and` condition for `Placeholder` name an see if the CI will pass tests.

**Comment by saikumarchalla** [comment]
Still issue exists in 2.5 and Please find the gist [here](https://colab.research.google.com/drive/1RuYXTvTCWpUiUk9di594_xS0gk0XgVWD#scrollTo=Pm15FYLLUbbg).Thanks!

**Comment by JakeTheSnake3p0** [comment]
In case this helps anyone, I had the same errors but for a different (buggy) reason. I created an [InputLayer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/InputLayer) and set the name to "Input" not thinking much of it. For whatever reason, that triggered this warning during save:

> WARNING:absl:Function _wrapped_model contains input name(s) a feature with unsupported characters which will be renamed to placeholder in the SavedModel.

Then when I tried loading the model I got the error:

> Unable to restore custom object of type _tf_keras_metric currently. Please make sure that the layer implements `get_config`and `from_config` when saving.

Even though I DID supply the custom_objects dict and saved using the SavedModel format. "Input" doesn't have unsupported characters but it triggered this issue anyway. Removing the name (or perhaps renaming to something else entirely?) avoided the problem but it doesn't seem to be a valid problem to begin with.
